File: go/5.go
/*

https://leetcode.com/problems/longest-palindromic-substring/

Given a string s, find the longest palindromic substring in s. You may assume that the maximum length of s is 1000.

Example 1:

Input: "babad"
Output: "bab"
Note: "aba" is also a valid answer.
Example 2:

Input: "cbbd"
Output: "bb"

*/

package main


/*
这个解法比较简单而且直接 但就是想不到啊
*/
func longestPalindrome(s string) string {
    res := ""
    for i := 0; i < len(s); i++ {
        left, right := i, i
        for ;left >= 0 && right < len(s); {
            if s[left] != s[right] {
                break
            }
            left--
            right++
        }
        if right - left - 1 > len(res) {
            res = s[left+1: right]
        }
        //println(i, left, right, res)
        if i + 1 < len(s) && s[i] == s[i+1] {
            left, right = i, i + 1
            for ;left >= 0 && right < len(s); {
                if s[left] != s[right] {
                    break
                }
                left--
                right++
            }
            if right - left - 1 > len(res) {
                res = s[left + 1: right]
            }
        }
    }
    return res
}


func main() {
    s := "babad"
    println(s, longestPalindrome(s))

    s = "aaa"
    println(s, longestPalindrome(s))

    s = ""
    println(s, longestPalindrome(s))

    s = "a"
    println(s, longestPalindrome(s))
}

File: dp/5.py
"""
https://leetcode.com/problems/longest-palindromic-substring/
需要注意很多边界条件
"""
class Solution:
    def longestPalindrome(self, s: str) -> str:
        res, N = "", len(s)
        dp = [[False] * N for _ in range(N)]
        # dp[i][j] = dp[i+1][j-1] if s[i] == s[j]
        for i in range(N-1, -1, -1):
            for j in range(i, N):
                # print(i, j, s[i:j+1])
                if i == j:
                    dp[i][j] = True
                elif i + 1 == j and s[i] == s[j]:
                    dp[i][j] = True
                else:
                    if s[i] == s[j]:
                        dp[i][j] |= dp[i+1][j-1]
                if dp[i][j]:
                    if j - i + 1 > len(res):
                        res = s[i: j+1]
        return res


if __name__ == "__main__":
    print(Solution().longestPalindrome("babad"))
    print(Solution().longestPalindrome("cbbd"))
